EpiserverRedirects.EntityFramework.Tests/Repository/RedirectRulesRepositoryTest.cs

```diff
@@ -0,0 +1,177 @@
+﻿using Forte.EpiserverRedirects.EntityFramework;
+using Forte.EpiserverRedirects.EntityFramework.Model;
+using Forte.EpiserverRedirects.EntityFramework.Repository;
+using Forte.EpiserverRedirects.Model.RedirectRule;
+using Forte.EpiserverRedirects.Repository;
+using Moq;
+using Moq.EntityFrameworkCore;
+using System;
+using System.Collections.Generic;
+using Xunit;
+
+
+namespace EpiserverRedirects.EntityFramework.Tests.Repository
+{
+    public class RedirectRulesRepositoryTest
+    {
+        private readonly IRedirectRuleRepository target;
+        private readonly Mock<IRedirectRulesDbContext> dbContext;
+        private readonly Mock<IRedirectRuleMapper> mapper;
+
+        private readonly Guid id1;
+        private readonly Guid id2;
+        private readonly Guid id3;
+        private readonly RedirectRuleEntity entity1;
+        private readonly RedirectRuleEntity entity2;
+        private readonly RedirectRuleEntity entity3;
+        private readonly RedirectRuleModel model1;
+        private readonly RedirectRuleModel model2;
+        private readonly RedirectRuleModel model3;
+
+        public RedirectRulesRepositoryTest()
+        {
+            dbContext = new Mock<IRedirectRulesDbContext>(MockBehavior.Strict);
+            mapper = new Mock<IRedirectRuleMapper>(MockBehavior.Strict);
+            target = new RedirectRulesRepository<IRedirectRulesDbContext>(dbContext.Object, mapper.Object);
+
+            id1 = Guid.NewGuid();
+            id2 = Guid.NewGuid();
+            id3 = Guid.NewGuid();
+            entity1 = new RedirectRuleEntity { Id = id1, OldPattern = "One", ContentId = 1111 };
+            entity2 = new RedirectRuleEntity { Id = id2, OldPattern = "Two", ContentId = 2222 };
+            entity3 = new RedirectRuleEntity { Id = id3, OldPattern = "Three", ContentId = null };
+            model1 = new RedirectRuleModel();
+            model2 = new RedirectRuleModel();
+            model3 = new RedirectRuleModel();
+            dbContext.Setup(o => o.RedirectRules).ReturnsDbSet(new List<RedirectRuleEntity> { entity1, entity2, entity3 });
+            mapper.Setup(o => o.MapToModel(entity1)).Returns(model1);
+            mapper.Setup(o => o.MapToModel(entity2)).Returns(model2);
+            mapper.Setup(o => o.MapToModel(entity3)).Returns(model3);
+        }
+
+        [Fact]
+        public void Given_Id_Returns_RedirectRule()
+        {
+            var actual = target.GetById(id2);
+            Assert.Same(model2, actual);
+        }
+
+        [Fact]
+        public void Given_GetAll_Returns_All()
+        {
+            var actual = target.GetAll();
+            Assert.Equal(3, actual.Count);
+            Assert.Same(model1, actual[0]);
```

Comment:
This is a white-box test, I know exactly what will be returned at each step by Mocked methods.
I strictly specified a list of 3 items instead of DbSet, and I strictly expect these 3 items in the output. Any change in order must result in a test failure.
If it was an integration test with some kind of test database - then yes, order is not important.

Reply:
But still, the order of these items is not the subject of this test. We should be able to reorder items in the initialization method without the need of fixing any tests.
It's better to always write such tests so that they are more resilient and they test only the "interesting" part of our code